[SR-4685] Can't infer associatedtypes with where clause

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-4685      |
   |Radar             | rdar://problem/39040799         |
   |Original Reporter | @rintaro      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Swift version 3.1-dev (LLVM 1067090d70, Clang 5e8cb0e694, Swift f83d1a70c7)  
Target: x86_64-unknown-linux-gnu


</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, TypeChecker        |
   |Assignee          | @DougGregor      |
   |Priority          | Medium      |

   

   md5: 3fae0cb45075895bc8535a198c649e47

  </details>




**is duplicated by**:  
* [SR-7304](https://bugs.swift.org/browse/SR-7304) typealias overriding associated type ... is better expressed as same-type constraint  


**Issue Description:**


The current master doesn't compile this.

``` java
protocol MySequence {
  associatedtype Element
  associatedtype Iterator: IteratorProtocol where Iterator.Element == Element
  func makeIterator() -> Iterator
}

struct Foo {}

struct FooIterator : IteratorProtocol {
  func next() -> Foo? { return Foo() }
}

struct FooSequence : MySequence {
  func makeIterator() -> FooIterator {
    return FooIterator()
  }
}
```

results:

``` java
test.swift:13:8: error: type 'FooSequence' does not conform to protocol 'MySequence'
struct FooSequence : MySequence {
       ^
test.swift:2:18: note: protocol requires nested type 'Element'; do you want to add it?
  associatedtype Element
                 ^
test.swift:3:18: note: protocol requires nested type 'Iterator'; do you want to add it?
  associatedtype Iterator: IteratorProtocol where Iterator.Element == Element
                 ^
```

This is basically the same structure as [PR-8939](https://github.com/apple/swift/pull/8939).

--- a/lib/Frontend/CompilerInvocation.cpp
+++ b/lib/Frontend/CompilerInvocation.cpp
@@ -1339,6 +1339,8 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
   if (Args.hasArg(OPT_warn_redundant_requirements))
     Opts.WarnRedundantRequirements = true;
 
+  Opts.EnableExperimentalAssociatedTypeInference = true;
+
   if (Args.hasArg(OPT_enable_experimental_associated_type_inference))
     Opts.EnableExperimentalAssociatedTypeInference = true;
   if (Args.hasArg(OPT_disable_experimental_associated_type_inference))

--- a/lib/Sema/AssociatedTypeInference.cpp
+++ b/lib/Sema/AssociatedTypeInference.cpp
@@ -2582,6 +2582,15 @@ bool AssociatedTypeInference::checkConstrainedExtension(ExtensionDecl *ext) {
   llvm_unreachable("unhandled result");
 }
 
+static bool containsConcreteDependentMemberType(Type ty) {
+  return ty.findIf([](Type t) -> bool {
+    if (auto *dmt = t->getAs<DependentMemberType>())
+      return !dmt->isTypeParameter();
+
+    return false;
+  });
+}
+
 AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
     ArrayRef<AssociatedTypeDecl *> unresolvedAssocTypes, unsigned reqDepth) {
   if (unresolvedAssocTypes.empty()) {
@@ -2709,15 +2718,15 @@ AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
       llvm::SmallPtrSet<AssociatedTypeDecl *, 4> circularityCheck;
       circularityCheck.insert(assocType);
 
-      std::function<Type(Type)> substCurrentTypeWitnesses;
-      substCurrentTypeWitnesses = [&](Type ty) -> Type {
+      std::function<llvm::Optional<Type>(Type)> substCurrentTypeWitnesses;
+      substCurrentTypeWitnesses = [&](Type ty) -> llvm::Optional<Type> {
         auto *const dmt = ty->getAs<DependentMemberType>();
         if (!dmt) {
-          return ty;
+          return llvm::None;
         }
 
         const auto substBase =
-            dmt->getBase().transform(substCurrentTypeWitnesses);
+            dmt->getBase().transformRec(substCurrentTypeWitnesses);
         if (!substBase) {
           return nullptr;
         }
@@ -2726,9 +2735,16 @@ AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
         // need to look up a tentative type witness. Otherwise, just substitute
         // the base.
         if (substBase->getAnyNominal() != dc->getSelfNominalTypeDecl()) {
-          return dmt->substBaseType(substBase,
-                                    LookUpConformanceInModule(dc->getParentModule()),
-                                    substOptions);
+          auto substTy = dmt->substBaseType(
+              substBase,
+              LookUpConformanceInModule(dc->getParentModule()),
+              substOptions);
+
+          // If any unresolved dependent member types remain, give up.
+          if (containsConcreteDependentMemberType(substTy))
+            return nullptr;
+
+          return substTy;
         }
 
         auto *assocTy = dmt->getAssocType();
@@ -2765,7 +2781,7 @@ AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
           // out so that we don't break any subst() invariants.
           if (tyWitness->hasTypeParameter() ||
               tyWitness->hasDependentMember()) {
-            tyWitness = tyWitness.transform(substCurrentTypeWitnesses);
+            tyWitness = tyWitness.transformRec(substCurrentTypeWitnesses);
           }
 
           if (tyWitness) {
@@ -2798,15 +2814,20 @@ AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
         return tyWitness;
       };
 
-      type = type.transform(substCurrentTypeWitnesses);
+      type = type.transformRec(substCurrentTypeWitnesses);
 
       // If substitution failed, give up.
-      if (!type || type->hasError())
+      if (!type || type->hasError()) {
+        LLVM_DEBUG(llvm::dbgs() << "-- Simplification failed\n");
         return assocType;
+      }
+
+      // If any unresolved dependent member types remain, give up.
+      assert(!containsConcreteDependentMemberType(type));
 
       type = dc->mapTypeIntoContext(type);
 
-      LLVM_DEBUG(llvm::dbgs() << "Substituted witness type is " << type << "\n";);
+      LLVM_DEBUG(llvm::dbgs() << "Simplified witness type is " << type << "\n";);
     }
 
     if (const auto failed =
@@ -4229,11 +4250,11 @@ TypeWitnessRequest::evaluate(Evaluator &eval,
           conformance, requirement);
 
       if (better == conformance) {
-        LLVM_DEBUG(llvm::dbgs() << "Conformance to " << conformance->getProtocol()
+        LLVM_DEBUG(llvm::dbgs() << "Conformance to " << conformance->getProtocol()->getName()
                                 << " is best\n";);
       } else {
-        LLVM_DEBUG(llvm::dbgs() << "Conformance to " << better->getProtocol()
-                                << " is better than " << conformance->getProtocol()
+        LLVM_DEBUG(llvm::dbgs() << "Conformance to " << better->getProtocol()->getName()
+                                << " is better than " << conformance->getProtocol()->getName()
                                 << "\n";);
       }
       if (better != conformance &&

--- a/lib/Sema/AssociatedTypeInference.cpp
+++ b/lib/Sema/AssociatedTypeInference.cpp
@@ -599,6 +599,8 @@ void InferredAssociatedTypesByWitness::dump(llvm::raw_ostream &out,
   out.indent(indent) << "(";
   if (Witness) {
     Witness->dumpRef(out);
+  } else {
+    out << "Tautological";
   }
 
   for (const auto &inferred : Inferred) {
@@ -948,12 +950,12 @@ private:
 
   /// Infer associated type witnesses for the given tentative
   /// requirement/witness match.
-  InferredAssociatedTypesByWitness inferTypeWitnessesViaValueWitness(
+  InferredAssociatedTypesByWitness getPotentialTypeWitnessesByMatchingTypes(
                                      ValueDecl *req,
                                      ValueDecl *witness);
 
   /// Infer associated type witnesses for the given value requirement.
-  InferredAssociatedTypesByWitnesses inferTypeWitnessesViaValueWitnesses(
+  InferredAssociatedTypesByWitnesses getPotentialTypeWitnessesFromRequirement(
                    const llvm::SetVector<AssociatedTypeDecl *> &allUnresolved,
                    ValueDecl *req);
 
@@ -1309,20 +1311,25 @@ static bool isExtensionUsableForInference(const ExtensionDecl *extension,
   if (!checkConformance(proto))
     return false;
 
-  // Source file and module file have different ways to get self bounds.
-  // Source file extension will have trailing where clause which can avoid
-  // computing a generic signature. Module file will not have
-  // trailing where clause, so it will compute generic signature to get
-  // self bounds which might result in slow performance.
+  // In a source file, we perform a syntactic check which avoids computing a
+  // generic signature. In a binary module, we have a generic signature so we
+  // can query it directly.
   SelfBounds bounds;
   if (extension->getParentSourceFile() != nullptr)
     bounds = getSelfBoundsFromWhereClause(extension);
-  else
+  else {
+    LLVM_DEBUG(llvm::dbgs() << "-- extension generic signature: "
+                            << extension->getGenericSignature() << "\n");
     bounds = getSelfBoundsFromGenericSignature(extension);
+  }
   for (auto *decl : bounds.decls) {
     if (auto *proto = dyn_cast<ProtocolDecl>(decl)) {
-      if (!checkConformance(proto))
+      if (!checkConformance(proto)) {
+        LLVM_DEBUG(llvm::dbgs() << "-- " << conformance->getType()
+                                << " does not conform to " << proto->getName()
+                                << "\n");
         return false;
+      }
     }
   }
 
@@ -1431,8 +1438,31 @@ static InferenceCandidateKind checkInferenceCandidate(
   return InferenceCandidateKind::Good;
 }
 
+/// Create an initial constraint system for the associated type inference solver.
+///
+/// Each protocol requirement defines a disjunction, where each disjunction
+/// element is a potential value witness for the requirement.
+///
+/// Each value witness binds some set of type witness candidates, which we
+/// compute by matching the witness type against the requirement.
+///
+/// The solver must pick exactly one value witness for each requirement while
+/// ensuring that the potential type bindings from each value witness are
+/// compatible with each other.
+///
+/// A value witness may be tautological, meaning it does not introduce any
+/// potential type witness bindings, for example, a protocol extension default
+/// `func f(_: Self.A) {}` for a protocol requirement `func f(_: Self.A)` with
+/// associated type A.
+///
+/// We collapse all tautological witnesses into one, since the solver only needs
+/// to explore that part of the solution space at most once. It is also true
+/// that it needs to explore it *at least* once, and we must take care when
+/// skipping potential bindings to distinguish between scenarios where a single
+/// binding is skipped, or the entire value witness must be thrown out because
+/// a binding is unsatisfiable.
 InferredAssociatedTypesByWitnesses
-AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
+AssociatedTypeInference::getPotentialTypeWitnessesFromRequirement(
                     const llvm::SetVector<AssociatedTypeDecl *> &allUnresolved,
                     ValueDecl *req) {
   // Conformances constructed by the ClangImporter should have explicit type
@@ -1448,6 +1478,9 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
   TypeReprCycleCheckWalker cycleCheck(dc->getASTContext(), allUnresolved);
 
   InferredAssociatedTypesByWitnesses result;
+
+  // Was there at least one witness that does not introduce new bindings?
+  bool hadTautologicalWitness = false;
 
   LLVM_DEBUG(llvm::dbgs() << "Considering requirement:\n";
              req->dump(llvm::dbgs()));
@@ -1473,7 +1506,7 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     }
 
     // Try to resolve the type witness via this value witness.
-    auto witnessResult = inferTypeWitnessesViaValueWitness(req, witness);
+    auto witnessResult = getPotentialTypeWitnessesByMatchingTypes(req, witness);
 
     // Filter out duplicated inferred types as well as inferred types
     // that don't meet the requirements placed on the associated type.
@@ -1493,12 +1526,16 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
       // Filter out errors.
       if (result.second->hasError()) {
         LLVM_DEBUG(llvm::dbgs() << "-- has error type\n");
+        // Skip this binding, but consider others from the same witness.
+        // This might not be strictly correct, but once we have error types
+        // we're diagnosing something anyway.
         REJECT;
       }
 
       // Filter out duplicates.
       if (!known.insert({result.first, result.second->getCanonicalType()})
                 .second) {
+        // Skip this binding, but consider others from the same witness.
         LLVM_DEBUG(llvm::dbgs() << "-- duplicate\n");
         REJECT;
       }
@@ -1517,18 +1554,20 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
 
       case InferenceCandidateKind::Tautological: {
         LLVM_DEBUG(llvm::dbgs() << "-- tautological\n");
+        // Skip this binding because it is immediately satisfied.
         REJECT;
       }
 
       case InferenceCandidateKind::Infinite: {
         LLVM_DEBUG(llvm::dbgs() << "-- infinite\n");
+        // Discard this witness altogether, because it has an unsatisfiable
+        // binding.
         goto next_witness;
       }
       }
 
-      // Check that the type witness doesn't contradict an
-      // explicitly-given type witness. If it does contradict, throw out the
-      // witness completely.
+      // Check that the binding doesn't contradict an explicitly-given type
+      // witness. If it does contradict, throw out the witness completely.
       if (!allUnresolved.count(result.first)) {
         auto existingWitness =
           conformance->getTypeWitness(result.first);
@@ -1560,6 +1599,10 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
           witnessResult.NonViable.push_back(
                           std::make_tuple(result.first,result.second,failed));
           LLVM_DEBUG(llvm::dbgs() << "-- doesn't fulfill requirements\n");
+
+          // By adding an element to NonViable we ensure the witness is rejected
+          // below, so we continue to consider other bindings to generate better
+          // diagnostics later.
           REJECT;
         }
       }
@@ -1569,9 +1612,13 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     }
 #undef REJECT
 
-    // If no inferred types remain, skip this witness.
-    if (witnessResult.Inferred.empty() && witnessResult.NonViable.empty())
+    // If no viable or non-viable bindings remain, the witness does not
+    // inter anything new, nor contradict any existing bindings. We collapse
+    // all tautological witnesses into a single element of the disjunction.
+    if (witnessResult.Inferred.empty() && witnessResult.NonViable.empty()) {
+      hadTautologicalWitness = true;
       continue;
+    }
 
     // If there were any non-viable inferred associated types, don't
     // infer anything from this witness.
@@ -1581,6 +1628,13 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     result.push_back(std::move(witnessResult));
 next_witness:;
 }
+
+  if (hadTautologicalWitness && !result.empty()) {
+    // Create a dummy entry, but only if there was at least one other witness;
+    // otherwise, we return an empty disjunction. See the remark in
+    // inferTypeWitnessesViaValueWitnesses() for explanation.
+    result.push_back(InferredAssociatedTypesByWitness());
+  }
 
   return result;
 }
@@ -1654,10 +1708,16 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
         continue;
     }
 
-    // Infer associated types from the potential value witnesses for
-    // this requirement.
+    // Collect this requirement's value witnesses and their potential
+    // type witness bindings.
     auto reqInferred =
-      inferTypeWitnessesViaValueWitnesses(assocTypes, req);
+      getPotentialTypeWitnessesFromRequirement(assocTypes, req);
+
+    // An empty disjunction is silently discarded, instead of immediately
+    // refuting the entirely system as it would in a real solver.
+    //
+    // If we find a solution and it so happens that this requirement cannot be
+    // witnessed, we'll diagnose the failure later in value witness checking.
     if (reqInferred.empty())
       continue;
 
@@ -1820,6 +1880,12 @@ AssociatedTypeInference::inferTypeWitnessesViaAssociatedType(
     result.push_back(std::move(inferred));
   }
 
+  if (!result.empty()) {
+    // If we found at least one default candidate, we must allow for the
+    // possibility that no default is chosen by adding a tautological witness
+    // to our disjunction.
+    result.push_back(InferredAssociatedTypesByWitness());
+  }
   return result;
 }
 
@@ -1878,10 +1944,11 @@ getReferencedAssocTypeOfProtocol(Type type, ProtocolDecl *proto) {
   return nullptr;
 }
 
-/// Attempt to resolve a type witness via a specific value witness.
+/// Find a set of potential type witness bindings by matching the interface type
+/// of the requirement against the interface type of a witness.
 InferredAssociatedTypesByWitness
-AssociatedTypeInference::inferTypeWitnessesViaValueWitness(ValueDecl *req,
-                                                           ValueDecl *witness) {
+AssociatedTypeInference::getPotentialTypeWitnessesByMatchingTypes(ValueDecl *req,
+                                                                  ValueDecl *witness) {
   InferredAssociatedTypesByWitness inferred;
   inferred.Witness = witness;
 
@@ -3032,16 +3099,28 @@ void AssociatedTypeInference::findSolutionsRec(
   for (const auto &witnessReq : inferredReq.second) {
     llvm::SaveAndRestore<unsigned> savedNumTypeWitnesses(numTypeWitnesses);
 
-    // If we inferred a type witness via a default, try both with and without
-    // the default.
-    if (isa<TypeDecl>(inferredReq.first)) {
-      // Recurse without considering this type.
+    // If we had at least one tautological witness, we must consider the
+    // possibility that none of the remaining witnesses are chosen.
+    if (witnessReq.Witness == nullptr) {
+      // Count tautological witnesses as if they come from protocol extensions,
+      // which ranks the solution lower than a more constrained one.
+      if (!isa<TypeDecl>(inferredReq.first))
+        ++numValueWitnessesInProtocolExtensions;
       valueWitnesses.push_back({inferredReq.first, nullptr});
       findSolutionsRec(unresolvedAssocTypes, solutions, nonViableSolutions,
                        valueWitnesses, numTypeWitnesses,
                        numValueWitnessesInProtocolExtensions, reqDepth + 1);
       valueWitnesses.pop_back();
+      if (!isa<TypeDecl>(inferredReq.first))
+        --numValueWitnessesInProtocolExtensions;
+      continue;
+    }
 
+    // If we inferred a type witness via a default, we do a slightly simpler
+    // thing.
+    //
+    // FIXME: Why can't we just fold this with the below?
+    if (isa<TypeDecl>(inferredReq.first)) {
       ++numTypeWitnesses;
       for (const auto &typeWitness : witnessReq.Inferred) {
         auto known = typeWitnesses.begin(typeWitness.first);
@@ -3075,6 +3154,7 @@ void AssociatedTypeInference::findSolutionsRec(
                llvm::dbgs() << " := ";
                witnessReq.Witness->dumpRef(llvm::dbgs());
                llvm::dbgs() << "\n";);
+
     valueWitnesses.push_back({inferredReq.first, witnessReq.Witness});
     if (!isa<TypeDecl>(inferredReq.first) &&
         witnessReq.Witness->getDeclContext()->getExtendedProtocolDecl())

--- a/lib/Sema/AssociatedTypeInference.cpp
+++ b/lib/Sema/AssociatedTypeInference.cpp
@@ -2956,6 +2956,14 @@ void AssociatedTypeInference::findSolutionsRec(
           unsigned numTypeWitnesses,
           unsigned numValueWitnessesInProtocolExtensions,
           unsigned reqDepth) {
+  // If this solution is going to be worse than what we've already recorded,
+  // give up now.
+  if (!solutions.empty() &&
+      solutions.front().NumValueWitnessesInProtocolExtensions
+          < numValueWitnessesInProtocolExtensions) {
+    return;
+  }
+
   using TypeWitnessesScope = decltype(typeWitnesses)::ScopeTy;
 
   // If we hit the last requirement, record and check this solution.
@@ -3088,6 +3096,13 @@ void AssociatedTypeInference::findSolutionsRec(
     solution.ValueWitnesses = valueWitnesses;
     solution.NumValueWitnessesInProtocolExtensions
       = numValueWitnessesInProtocolExtensions;
+
+    // If this solution was clearly better than the previous best solution,
+    // swap them.
+    if (solutionList.back().NumValueWitnessesInProtocolExtensions
+          < solutionList.front().NumValueWitnessesInProtocolExtensions) {
+      std::swap(solutionList.front(), solutionList.back());
+    }
 
     // We're done recording the solution.
     return;
@@ -3441,15 +3456,10 @@ bool AssociatedTypeInference::findBestSolution(
   if (solutions.empty()) return true;
   if (solutions.size() == 1) return false;
 
-  // Find the smallest number of value witnesses found in protocol extensions.
-  // FIXME: This is a silly heuristic that should go away.
+  // The solution at the front has the smallest number of value witnesses found
+  // in protocol extensions, by construction.
   unsigned bestNumValueWitnessesInProtocolExtensions
     = solutions.front().NumValueWitnessesInProtocolExtensions;
-  for (unsigned i = 1, n = solutions.size(); i != n; ++i) {
-    bestNumValueWitnessesInProtocolExtensions
-      = std::min(bestNumValueWitnessesInProtocolExtensions,
-                 solutions[i].NumValueWitnessesInProtocolExtensions);
-  }
 
   // Erase any solutions with more value witnesses in protocol
   // extensions than the best.

--- a/test/Generics/associated_type_where_clause.swift
+++ b/test/Generics/associated_type_where_clause.swift
@@ -120,14 +120,14 @@ struct ConcreteInheritsDiffer: Inherits {
     typealias U = ConcreteConforms
     typealias X = ConcreteConforms2
 }
-/*
-FIXME: the sametype requirement gets dropped from the requirement signature
-(enumerateRequirements doesn't yield it), so this doesn't error as it should.
+
 struct BadConcreteInherits: Inherits {
+// expected-error@-1 {{type 'BadConcreteInherits' does not conform to protocol 'Inherits'}}
+// expected-error@-2 {{'Inherits' requires the types 'ConcreteConforms.T' (aka 'Int') and 'ConcreteConformsNonFoo2.T' (aka 'Float') be equivalent}}
+// expected-note@-3 {{requirement specified as 'Self.U.T' == 'Self.X.T' [with Self = BadConcreteInherits]}}
     typealias U = ConcreteConforms
     typealias X = ConcreteConformsNonFoo2
 }
-*/
 
 struct X { }
 

--- a/test/decl/protocol/req/associated_type_inference_exponential.swift
+++ b/test/decl/protocol/req/associated_type_inference_exponential.swift
@@ -1,0 +1,145 @@
+// RUN: %target-typecheck-verify-swift
+
+// Inspired by GRDB.swift.
+
+protocol P {
+  associatedtype A
+  associatedtype B
+  func f1(_: A)
+  func f2(_: A)
+  func f3(_: A)
+  func f4(_: A)
+  func f5(_: A)
+  func f6(_: A)
+  func f7(_: A)
+  func f8(_: A)
+  func f9(_: A)
+  func f10(_: A)
+  func f11(_: A)
+  func f12(_: A)
+  func f13(_: A)
+  func f14(_: A)
+  func f15(_: A)
+  func f16(_: B)
+  func f17(_: B)
+  func f18(_: B)
+  func f19(_: B)
+  func f20(_: B)
+  func f21(_: B)
+  func f22(_: B)
+  func f23(_: B)
+  func f24(_: B)
+  func f25(_: B)
+  func f26(_: B)
+  func f27(_: B)
+  func f28(_: B)
+  func f29(_: B)
+  func f30(_: B)
+  func f31(_: B)
+  func f32(_: B)
+}
+
+extension P {
+  func f1(_: A) {}
+  func f2(_: A) {}
+  func f3(_: A) {}
+  func f4(_: A) {}
+  func f5(_: A) {}
+  func f6(_: A) {}
+  func f7(_: A) {}
+  func f8(_: A) {}
+  func f9(_: A) {}
+  func f10(_: A) {}
+  func f11(_: A) {}
+  func f12(_: A) {}
+  func f13(_: A) {}
+  func f14(_: A) {}
+  func f15(_: A) {}
+  func f16(_: A) {}
+  func f17(_: A) {}
+  func f18(_: A) {}
+  func f19(_: A) {}
+  func f20(_: A) {}
+  func f21(_: A) {}
+  func f22(_: A) {}
+  func f23(_: A) {}
+  func f24(_: A) {}
+  func f25(_: A) {}
+  func f26(_: A) {}
+  func f27(_: A) {}
+  func f28(_: A) {}
+  func f29(_: A) {}
+  func f30(_: A) {}
+  func f31(_: A) {}
+  func f32(_: A) {}
+}
+
+extension P {
+  func f1(_: B) {}
+  func f2(_: B) {}
+  func f3(_: B) {}
+  func f4(_: B) {}
+  func f5(_: B) {}
+  func f6(_: B) {}
+  func f7(_: B) {}
+  func f8(_: B) {}
+  func f9(_: B) {}
+  func f10(_: B) {}
+  func f11(_: B) {}
+  func f12(_: B) {}
+  func f13(_: B) {}
+  func f14(_: B) {}
+  func f15(_: B) {}
+  func f16(_: B) {}
+  func f17(_: B) {}
+  func f18(_: B) {}
+  func f19(_: B) {}
+  func f20(_: B) {}
+  func f21(_: B) {}
+  func f22(_: B) {}
+  func f23(_: B) {}
+  func f24(_: B) {}
+  func f25(_: B) {}
+  func f26(_: B) {}
+  func f27(_: B) {}
+  func f28(_: B) {}
+  func f29(_: B) {}
+  func f30(_: B) {}
+  func f31(_: B) {}
+  func f32(_: B) {}
+}
+
+struct S: P {
+  func f1(_: Int) {}
+  func f2(_: Int) {}
+  func f3(_: Int) {}
+  func f4(_: Int) {}
+  func f5(_: Int) {}
+  func f6(_: Int) {}
+  func f7(_: Int) {}
+  func f8(_: Int) {}
+  func f9(_: Int) {}
+  func f10(_: Int) {}
+  func f11(_: Int) {}
+  func f12(_: Int) {}
+  func f13(_: Int) {}
+  func f14(_: Int) {}
+  func f15(_: Int) {}
+  func f16(_: Int) {}
+  func f17(_: Int) {}
+  func f18(_: Int) {}
+  func f19(_: Int) {}
+  func f20(_: Int) {}
+  func f21(_: Int) {}
+  func f22(_: Int) {}
+  func f23(_: Int) {}
+  func f24(_: Int) {}
+  func f25(_: Int) {}
+  func f26(_: Int) {}
+  func f27(_: Int) {}
+  func f28(_: Int) {}
+  func f29(_: Int) {}
+  func f30(_: Int) {}
+  func f31(_: Int) {}
+  func f32(_: Int) {}
+}

--- a/test/decl/protocol/req/associated_type_inference_stdlib_5.swift
+++ b/test/decl/protocol/req/associated_type_inference_stdlib_5.swift
@@ -1,0 +1,55 @@
+// RUN: %target-swift-frontend -emit-silgen %s
+
+// Reduced from swift-nio.
+
+public protocol AppendableCollection: Collection {
+    mutating func append(_ newElement: Self.Iterator.Element)
+}
+
+public struct CircularBuffer<E>: AppendableCollection {
+    public mutating func append(_ value: E) {
+        fatalError()
+    }
+
+    private mutating func doubleCapacity() {
+        fatalError()
+    }
+
+    public subscript(index: Int) -> E {
+        get {
+            fatalError()
+        }
+        set {
+            fatalError()
+        }
+    }
+
+    public var indices: Range<Int> {
+        fatalError()
+    }
+
+    public var isEmpty: Bool {
+        fatalError()
+    }
+
+    public var count: Int {
+        fatalError()
+    }
+
+    public var startIndex: Int {
+        fatalError()
+    }
+
+    public var endIndex: Int {
+        fatalError()
+    }
+
+    public func index(after: Int) -> Int {
+        fatalError()
+    }
+
+    public func index(before: Int) -> Int {
+        fatalError()
+    }
+}
+

--- a/test/decl/protocol/req/associated_type_inference_tautology.swift
+++ b/test/decl/protocol/req/associated_type_inference_tautology.swift
@@ -1,6 +1,8 @@
 // RUN: %target-typecheck-verify-swift -disable-experimental-associated-type-inference
 // RUN: %target-typecheck-verify-swift -enable-experimental-associated-type-inference
 
+// A := G<A> is unsatisfiable and not a tautology!
+
 protocol P1 {
   associatedtype A
   associatedtype B
@@ -26,6 +28,8 @@ struct S1: P1 {}
 let x1: String.Type = S1.A.self
 let y1: String.Type = S1.B.self
 
+// Potential witness has innermost generic parameters.
+
 protocol P2 {
   associatedtype A
   associatedtype B
@@ -48,3 +52,26 @@ struct S2: P2 {}
 
 let x2: String.Type = S2.A.self
 let y2: String.Type = S2.B.self
+
+// If all type witness bindings were tautological, we must still consider the
+// witness as introducing no bindings.
+
+protocol P3 {
+  associatedtype A
+
+  func f(_: A)
+  func g(_: A)
+}
+
+extension P3 {
+  func g(_: A) {}
+
+  // We should not be forced to choose g().
+  func g(_: String) {}
+}
+
+struct S3: P3 {
+  func f(_: Int) {}
+}
+
+let x3: Int.Type = S3.A.self

